### ISSUE TITLE
Deployment on custom IP, problem with worker connection

### ISSUE THREAD

**Issue by MrStrzelec** [mljar/mercury#391]
Hi there!
First of all I'd like to say that I really appreciate your software. 
But I'm having trouble with a worker.

Basics:
We don't use any kind of huggingface/docker, so it only works on localhost. 
We've run everything as u said in the tutorial, but we're not able to run it correctly. 

So we're running server via systemd -  python manage.py runserver <IP ADDRESS>.
We'll also run worker as systemd -  celery -A server worker --loglevel=info -P gevent --concurrency 1 -E
(to run the mercury server no matter is server is restarted)

And yet we're still waiting for Worker.
![Mercury Worker](https://github.com/mljar/mercury/assets/29755438/8f8f38b9-65f6-400d-8c1f-fe46122ce0f4)

Do you have any idea what's going on?
Thank you very much!

**Comment by pplonski** [comment]
Hi @MrStrzelec,

Have you tried to start with `mercury run`? Additionally you can start server in verbose mode `mercury run --verbose`. 

If you would like to run each part of architecture by your own I would recommend checking docker entrypoint script: https://github.com/mljar/mercury/blob/main/docker/mercury/entrypoint.sh 

In your case, looks like you are missing: `-Q celery,ws`, please try to run `celery -A server worker --loglevel=info -P gevent --concurrency 1 -E -Q celery,ws`

Please let me know if it works for you.

**Comment by MrStrzelec** [comment]
Hey!

Server working. 
I've managed also to change celery script and adjust to your advice but not working. I've got this error from celery (systemd).

Okt 30 09:22:45 Servername celery[1155]:     return request("get", url, params=params, **kwargs)
Okt 30 09:22:45 Servername celery[1155]:   File "/opt/jupyterhub/lib/python3.10/site-packages/requests/api.py", line 59, in request
Okt 30 09:22:45 Servername celery[1155]:     return session.request(method=method, url=url, **kwargs)
Okt 30 09:22:45 Servername celery[1155]:   File "/opt/jupyterhub/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
Okt 30 09:22:45 Servername celery[1155]:     resp = self.send(prep, **send_kwargs)
Okt 30 09:22:45 Servername celery[1155]:   File "/opt/jupyterhub/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
Okt 30 09:22:45 Servername celery[1155]:     r = adapter.send(request, **kwargs)
Okt 30 09:22:45 Servername celery[1155]:   File "/opt/jupyterhub/lib/python3.10/site-packages/requests/adapters.py", line 519, in send
Okt 30 09:22:45 Servername celery[1155]:     raise ConnectionError(e, request=request)
Okt 30 09:22:45 Servername celery[1155]: requests.exceptions.ConnectionError: HTTPConnectionPool(host=Servername', port=5668): Max retries exceeded with url:  /api/v1/worker/18db8815-183d-4f84-929b-40522256a3e2/387/12/nb (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb144005870>:Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by pplonski** [comment]
I have no idea right now what might be the issue. Is it running on 127.0.0.1?

**Comment by MrStrzelec** [comment]
It's running on a diffrent IP address. I thought i add this information.

python manage.py runserver <IPADDRESS:PORT>

**Comment by pplonski** [comment]
Could you please add `--verbose` argument to server start command? Let's check if worker is connecting to the correct IP address.

**Comment by MrStrzelec** [comment]
user@server:/opt/jupyterhub/lib/python3.10/site-packages/mercury$ python manage.py runsever --verbose IP:PORT
Traceback (most recent call last):
  File "/usr/lib/python3.10/logging/config.py", line 565, in configure
    handler = self.configure_handler(handlers[name])
  File "/usr/lib/python3.10/logging/config.py", line 746, in configure_handler
    result = factory(**kwargs)
  File "/usr/lib/python3.10/logging/__init__.py", line 1169, in __init__
    StreamHandler.__init__(self, self._open())
  File "/usr/lib/python3.10/logging/__init__.py", line 1201, in _open
    return open_func(self.baseFilename, self.mode,
PermissionError: [Errno 13] Permission denied: '/opt/jupyterhub/lib/python3.10/site-packages/mercury/django-errors.log'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/jupyterhub/lib/python3.10/site-packages/mercury/manage.py", line 22, in <module>
    main()
  File "/opt/jupyterhub/lib/python3.10/site-packages/mercury/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/opt/jupyterhub/lib/python3.10/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/opt/jupyterhub/lib/python3.10/site-packages/django/core/management/__init__.py", line 416, in execute
    django.setup()
  File "/opt/jupyterhub/lib/python3.10/site-packages/django/__init__.py", line 19, in setup
    configure_logging(settings.LOGGING_CONFIG, settings.LOGGING)
  File "/opt/jupyterhub/lib/python3.10/site-packages/django/utils/log.py", line 76, in configure_logging
    logging_config_func(logging_settings)
  File "/usr/lib/python3.10/logging/config.py", line 811, in dictConfig
    dictConfigClass(config).configure()
  File "/usr/lib/python3.10/logging/config.py", line 572, in configure
    raise ValueError('Unable to configure handler '
ValueError: Unable to configure handler 'file'

It tried to run also mercury run --verbose IP:PORT and output is similar

**Comment by pplonski** [comment]
Looks like process that is running server doesnt have permission to write logs to the directory. This might be connected to https://github.com/mljar/mercury/issues/384 I will need some time to make it configurable.

Have you already developed the notebook with mercury and deployment is the last step?

**Comment by MrStrzelec** [comment]
`PermissionError: [Errno 13] Permission denied: '/opt/jupyterhub/lib/python3.10/site-packages/mercury/django-errors.log'`

I've just changed permissions to this file for everyone and now it's works. Server running correctly and worker working corretly. W/O docker or huggingface. For some reasons Mercury showing that he cannot find this file, while JupyterNotebook working correctly. U know why?

![File not found](https://github.com/mljar/mercury/assets/29755438/1485b0e9-5c46-42ea-b08f-d4a2feaf4546)

**Comment by pplonski** [comment]
Great that it is working!

Do you have data file in the same directory as notebook? It should read file if it is in the same directory.

Alternative might be to provide full path to the file.

**Comment by MrStrzelec** [comment]
Hey!

Sorry i had a day off yesterday. 

Yep, data is in the same directory as a notebook. I also tried with full path and also didn't worked. 

Any other idea?
Thanks a Ton :)

**Comment by pplonski** [comment]
This is strange, could you provide full error message? Maybe try to print out the working directory before loading files, just to check the path. Try to add code:

```python
import os
print(os.getcwd())
```

**Comment by MrStrzelec** [comment]
I think worker was not working at all, i've checked logs and found this:

`[2023-11-02 12:33:20,930: INFO/MainProcess] Task apps.ws.tasks.task_start_websocket_worker[1cfbef57-5ef1-4ea6-b307-f1380f3a20aa] received
[2023-11-02 12:33:20,976: INFO/MainProcess] Task apps.ws.tasks.task_start_websocket_worker[1cfbef57-5ef1-4ea6-b307-f1380f3a20aa] succeeded in 0.04001394798979163s: None
NB 2023-11-02 12:33:22,661 Exception when notebook load, quit
Traceback (most recent call last):
  File "/opt/jupyterhub/lib/python3.10/site-packages/mercury/apps/../apps/nbworker/rest.py", line 36, in load_notebook
    raise Exception("Cant load notebook")
Exception: Cant load notebook
`

I think worker wouldn't find this file due to having problem with loading notebook.
I've check everything twice and didn't find anything.

**Comment by pplonski** [comment]
Do you have notebook file in the same directory in which you start worker?

**Comment by keyvan-najafy** [comment]
Hi
I am having the same problem

I have created a private fork without any changes and deployed for http via docker-compose on my local computer with default setting and changed nothing but I am getting similar errors to this https://github.com/mljar/mercury/issues/391#issuecomment-1784703244

I have used [mercury-deploy-demo repo](https://github.com/mljar/mercury-deploy-demo) for notebooks and everything seems to be working. admin panel, welcome.md , and notebook icons in welcome page are getting displayed properly but when I click on any of notebooks I get following errors in terminal and I am redirected to notebooks page and notebook is visible in the background but page stays in loading state forever

what should I do ?


    mercuryprivate-mercury-1  |   NB 2023-11-23 22:43:06,960 Exception when notebook load, quit
    mercuryprivate-mercury-1  |   Traceback (most recent call last):
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connection.py", line 174, in _new_conn
    mercuryprivate-mercury-1  |     conn = connection.create_connection(
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/util/connection.py", line 95, in create_connection
    mercuryprivate-mercury-1  |     raise err
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/util/connection.py", line 85, in create_connection
    mercuryprivate-mercury-1  |     sock.connect(sa)
    mercuryprivate-mercury-1  | ConnectionRefusedError: [Errno 111] Connection refused
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | During handling of the above exception, another exception occurred:
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | Traceback (most recent call last):
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connectionpool.py", line 715, in urlopen
    mercuryprivate-mercury-1  |     httplib_response = self._make_request(
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connectionpool.py", line 416, in _make_request
    mercuryprivate-mercury-1  |     conn.request(method, url, **httplib_request_kw)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connection.py", line 244, in request
    mercuryprivate-mercury-1  |     super(HTTPConnection, self).request(method, url, body=body, headers=headers)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/http/client.py", line 1283, in request
    mercuryprivate-mercury-1  |     self._send_request(method, url, body, headers, encode_chunked)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/http/client.py", line 1329, in _send_request
    mercuryprivate-mercury-1  |     self.endheaders(body, encode_chunked=encode_chunked)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/http/client.py", line 1278, in endheaders
    mercuryprivate-mercury-1  |     self._send_output(message_body, encode_chunked=encode_chunked)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/http/client.py", line 1038, in _send_output
    mercuryprivate-mercury-1  |     self.send(msg)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/http/client.py", line 976, in send
    mercuryprivate-mercury-1  |     self.connect()
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connection.py", line 205, in connect
    mercuryprivate-mercury-1  |     conn = self._new_conn()
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connection.py", line 186, in _new_conn
    mercuryprivate-mercury-1  |     raise NewConnectionError(
    mercuryprivate-mercury-1  | urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f16bc30e3e0>: Failed to establish a new connection: [Errno 111] Connection refused
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | During handling of the above exception, another exception occurred:
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | Traceback (most recent call last):
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/adapters.py", line 486, in send
    mercuryprivate-mercury-1  |     resp = conn.urlopen(
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/connectionpool.py", line 799, in urlopen
    mercuryprivate-mercury-1  |     retries = retries.increment(
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/urllib3/util/retry.py", line 592, in increment
    mercuryprivate-mercury-1  |     raise MaxRetryError(_pool, url, error or ResponseError(cause))
    mercuryprivate-mercury-1  | urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=80): Max retries exceeded with url: /api/v1/worker/dd83117b-dcd9-4c54-8782-aa60c094f6ac/16/1/nb (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f16bc30e3e0>: Failed to establish a new connection: [Errno 111] Connection refused'))
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | During handling of the above exception, another exception occurred:
    mercuryprivate-mercury-1  |
    mercuryprivate-mercury-1  | Traceback (most recent call last):
    mercuryprivate-mercury-1  |   File "/app/mercury/apps/../apps/nbworker/rest.py", line 32, in load_notebook
    mercuryprivate-mercury-1  |     response = requests.get(
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/api.py", line 73, in get
    mercuryprivate-mercury-1  |     return request("get", url, params=params, **kwargs)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/api.py", line 59, in request
    mercuryprivate-mercury-1  |     return session.request(method=method, url=url, **kwargs)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
    mercuryprivate-mercury-1  |     resp = self.send(prep, **send_kwargs)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
    mercuryprivate-mercury-1  |     r = adapter.send(request, **kwargs)
    mercuryprivate-mercury-1  |   File "/opt/conda/lib/python3.10/site-packages/requests/adapters.py", line 519, in send
    mercuryprivate-mercury-1  |     raise ConnectionError(e, request=request)
    mercuryprivate-mercury-1  | requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=80): Max retries exceeded with url: /api/v1/worker/dd83117b-dcd9-4c54-8782-aa60c094f6ac/16/1/nb (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f16bc30e3e0>: Failed to establish a new connection: [Errno 111] Connection refused'))

**Comment by kskadart** [comment]
Hello folks, face the same issue like @keyvan-najafy

**Comment by pplonski** [comment]
Hi @keyvan-najafy, @kskadart,

What operating system are you using? How to reproduce your environment?

**Comment by keyvan-najafy** [comment]
@pplonski adding REACT_APP_SERVER_WS did not help.
I managed to fix it somehow by editing mercury/apps/nbworker/__main__.py
The default server url is ws://localhost which is given to __main__.py by enterypoint.sh (I think so no sure!!)
I simply changed this to ws://localhost:9000 and everything started to working
```
if __name__ == "__main__":
    log.info(f"Start NBWorker with arguments {sys.argv}")
    server_url = 'ws://localhost:9000' #### added this line
    for _ in range(CONNECT_MAX_TRIES):
        nb_worker = NBWorker(
            f"{server_url}/ws/worker/{notebook_id}/{session_id}/{worker_id}/",
            notebook_id,
            session_id,
            worker_id,
        )
        time.sleep(RECONNECT_WAIT_TIME)
```
so basicly the idea is that somehow either by enterypoint.sh or docker-compose.yml there is a mismatching for ports given to the mercury container
to sum up everything I have done:
- add MERCURY_SERVER_URL: ${MERCURY_SERVER_URL} to docker-compose file under environment section
- add MERCURY_SERVER_URL=http://127.0.0.1:9000 to .env file
- add ```server_url = 'ws://localhost:9000' ``` to mercury/apps/nbworker/__main__.py under ```if __name__ == "__main__":```

This is just a temporary fix as it is pretty bad practice to change `__main.py__ `directly so perhaps enterypoint.sh or docker-compose.yml should get an update perhaps even creating a new issue.

I am planning to deploy this on a local cloud platform which does not support docker compose (so should put all containers separately). I hope these fixes don't not create anymore complications

**Comment by pplonski** [comment]
Thank you @keyvan-najafy for steps description with fix. Yes, there is some miss mach that should be fixed.

**Comment by ghansham** [comment]
@MrStrzelec, did you get success. I am also getting exactly same error and running on jupyterhub. The web socket connection is successful but worker is queued and getting same error in rest.py in load_notebook method at line no. 32.

**Comment by mariliaribeiro** [comment]
> @pplonski adding REACT_APP_SERVER_WS did not help. I managed to fix it somehow by editing mercury/apps/nbworker/**main**.py The default server url is ws://localhost which is given to **main**.py by enterypoint.sh (I think so!!) I simply changed this to ws://localhost:9000 and everything started to working
> 
> ```
> if __name__ == "__main__":
>     log.info(f"Start NBWorker with arguments {sys.argv}")
>     server_url = 'ws://localhost:9000' #### added this line
>     for _ in range(CONNECT_MAX_TRIES):
>         nb_worker = NBWorker(
>             f"{server_url}/ws/worker/{notebook_id}/{session_id}/{worker_id}/",
>             notebook_id,
>             session_id,
>             worker_id,
>         )
>         time.sleep(RECONNECT_WAIT_TIME)
> ```
> 
> so basicly the idea is that somehow either by enterypoint.sh or docker-compose.yml there is a mismatching for ports given to the mercury container to sum up everything I have done:
> 
> * add MERCURY_SERVER_URL: ${MERCURY_SERVER_URL} to docker-compose file under environment section
> * add MERCURY_SERVER_URL=http://127.0.0.1:9000 to .env file
> * add `server_url = 'ws://localhost:9000' ` to mercury/apps/nbworker/**main**.py under `if __name__ == "__main__":`
> 
> This is just a temporary fix as it is pretty bad practice to change `__main.py__ `directly so perhaps enterypoint.sh or docker-compose.yml should get an update or even creating a new issue.
> 
> I am planning to deploy this on a local cloud platform which does not support docker compose (so I should put all containers separately). I hope these fixes don't not create anymore complications

Thanks, it's working for me!

I think the main problem is because the server port is not present in the sys.argv called in `mercury/apps/nbworker/main.py`. The log is showing:

**Comment by pplonski** [comment]
Hi @mariliaribeiro,

Here is code that creates job for worker https://github.com/mljar/mercury/blob/4a404b30cb601275c1f591c6d9f9ef22b584698b/mercury/apps/ws/client.py#L137-L148

Here is code that starts worker process based on job params
https://github.com/mljar/mercury/blob/4a404b30cb601275c1f591c6d9f9ef22b584698b/mercury/apps/ws/tasks.py#L34-L43

Sorry, that not everything is working smoothly ... I'm focused right now on ipywidgets support in Mercury. 

Anyway, have you successfully created app in Mercury? What is your use case? Do you enjoy working with Mercury?

**Comment by nagatushar** [comment]
> `PermissionError: [Errno 13] Permission denied: '/opt/jupyterhub/lib/python3.10/site-packages/mercury/django-errors.log'`
> 
> I've just changed permissions to this file for everyone and now it's works. Server running correctly and worker working correctly. W/O docker or huggingface. For some reasons Mercury showing that he cannot find this file, while JupyterNotebook working correctly. U know why?
> 
> ![File not found](https://private-user-images.githubusercontent.com/29755438/279049639-1485b0e9-5c46-42ea-b08f-d4a2feaf4546.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4NTg1OTIsIm5iZiI6MTcwNzg1ODI5MiwicGF0aCI6Ii8yOTc1NTQzOC8yNzkwNDk2MzktMTQ4NWIwZTktNWM0Ni00MmVhLWIwOGYtZDRhMmZlYWY0NTQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEzVDIxMDQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmM2QxN2RlODJiZWZlZTM1ZDczMWZhNTljZWMxMjE1MjZmMjEwOWQ1NmViNjQ3NjllMzkzMzM5YzA0M2ExMjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nsyQoPhGzz77UpLPrSP71CXgYgIU8Vypz3BtQIky9uE)

/app/ "your file path"

**Comment by somdipdatta** [comment]
First: I love how easy it is to build apps from notebook, using mercury.

The problem I am having is likely related to this thread, hence posting here.
I deployed my mercury app in render.com, they do not allow 127.0.0.0, hence I am running
`mercury run 0.0.0.0:8000`

The client shows "Waiting for Worker.." and never connects.
I tried setting env `REACT_APP_SERVER_WS` and `MERCURY_SERVER_URL`, neither helped.

**Comment by pplonski** [comment]
This might be an issue with hard coded worker address.

**Comment by ghansham** [comment]
Yes it is. We had to modify that

**Comment by somdipdatta** [comment]
> Yes it is. We had to modify that

modified in an upcoming release?

**Comment by AGrzes** [comment]
I'm facing the same issue.

My understanding is that the root cause is having single `server_url` / `server_address`  that is used by both front-end and nbworker while
* front-end may go through arbitrary number of reverse-proxies, load balancers, have ssl termination in front and so on
* nbworker should go through shortest possible path -loopback / direct machine to machine / singe load balancer

at the moment the method to determine first address determines second address - and simple solution would be to allow overriding it for deployments that need it.

Now - the issue is marked `help wanted` - would PR be welcome?

**Comment by AGrzes** [comment]
I think the PR https://github.com/mljar/mercury/pulls address exactly this. 

Maybe the name `MERCURY_TASK_SERVER_URL` is not clear but it is "The address nbworker processes used to connect to main server through websockets"
